swupd/manifest.go

```diff
@@ -317,9 +331,15 @@ func (m *Manifest) sortFilesName() {
 	sort.Slice(m.DeletedFiles, func(i, j int) bool {
 		return m.DeletedFiles[i].Name < m.DeletedFiles[j].Name
 	})
+	m.Sorted = sortedByName
 }
 
 func (m *Manifest) sortFilesVersionName() {
+	// do not re-sort
+	if m.Sorted == sortedByVersionName {
+		return
+	}
```

Comment:
I don't like this solution because it will hide from us when the expensive operation is happening, and led us to write code that is going to be more care to analyze. Might sound trivial know to understand, but put a few months of patches on top, and things become difficult.

Given we have a good idea of when things need to be sorted, _my vote is to make sorting more explicit_.

- We read all the bundle manifest files
- Sort them by name
- Run other algorithms that depend on sorted manifest
 -- Link Peers, Find Gone, Maximize Full
 -- Those that change the list, should fix it before return
- In the end, sort them by version then name

Easy fix for algorithms that change the list is to call Sort before they return, but this approach let us try other ways to keep the invariant. E.g. Find Gone can first store the deleted files list, and then create a new one (the right size) merging both Files and DeletedFiles, then just doing m.Files = mergedFiles.

That said, if you still want to pursue the approach in this patch, make a new type that is a struct containing the list of files and the sorted state, and create methods to operate on that. You lose the benefit of using the x[y] syntax, but you properly encapsulate and it will be clearer that we shouldn't be appending to it, but calling m.Files.appendFile().